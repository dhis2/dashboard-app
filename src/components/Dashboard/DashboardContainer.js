import React from 'react'
import PropTypes from 'prop-types'
import cx from 'classnames'

import classes from './styles/DashboardContainer.module.css'

const DashboardContainer = ({ children, covered }) => {
    return (
        <div
            className={cx(
                classes.container,
                'dashboard-scroll-container',
                covered && classes.covered
            )}
        >
            {children}
        </div>
    )
}

DashboardContainer.propTypes = {
    children: PropTypes.node,
    covered: PropTypes.bool,
}

export default DashboardContainer
