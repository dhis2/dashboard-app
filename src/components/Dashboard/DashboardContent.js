import React, { Fragment } from 'react';
import TitleBar from '../TitleBar/TitleBar';
import ItemGrid from '../ItemGrid/ItemGrid';

export const DashboardContent = props => (
    <Fragment>
        <TitleBar edit={props.editMode} />
        <ItemGrid edit={props.editMode} />
    </Fragment>
);

export default DashboardContent;
