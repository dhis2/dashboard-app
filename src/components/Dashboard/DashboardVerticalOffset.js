import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import {
    getControlBarHeight,
    MIN_ROW_COUNT,
} from '../ControlBar/controlBarDimensions';
import { sGetControlBarUserRows } from '../../reducers/controlBar';

const DashboardVerticalOffset = props => {
    const rows = props.editMode ? MIN_ROW_COUNT : props.userRows;
    const marginTop = getControlBarHeight(rows);

    return <div className="page-container-top-margin" style={{ marginTop }} />;
};

const mapStateToProps = state => ({
    userRows: sGetControlBarUserRows(state),
});

DashboardVerticalOffset.propTypes = {
    editMode: PropTypes.bool,
    userRows: PropTypes.number,
};

export default connect(mapStateToProps)(DashboardVerticalOffset);
