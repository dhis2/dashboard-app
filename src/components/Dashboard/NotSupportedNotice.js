import React from 'react'
import PropTypes from 'prop-types'
import i18n from '@dhis2/d2-i18n'
import { NoticeBox, CenteredContent } from '@dhis2/ui'

import classes from './styles/NotSupportedNotice.module.css'

const NotSupportedNotice = ({ message }) => (
    <CenteredContent position="top">
        <div className={classes.noticeContainer}>
            <NoticeBox title={i18n.t('Not supported')} warning>
                {message}
            </NoticeBox>
        </div>
    </CenteredContent>
)

NotSupportedNotice.propTypes = {
    message: PropTypes.string,
}

export default NotSupportedNotice
