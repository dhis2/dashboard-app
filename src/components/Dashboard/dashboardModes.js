export const EDIT = 'edit';
export const NEW = 'new';
export const VIEW = 'view';
