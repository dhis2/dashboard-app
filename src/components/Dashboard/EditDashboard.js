import React, { Component } from 'react'
import { connect } from 'react-redux'
import i18n from '@dhis2/d2-i18n'
import PropTypes from 'prop-types'

import { acSetEditDashboard } from '../../actions/editDashboard'
import { sGetSelectedId } from '../../reducers/selected'
import {
    sGetDashboardById,
    sGetDashboardItems,
    sDashboardsIsFetching,
} from '../../reducers/dashboards'
import DashboardVerticalOffset from './DashboardVerticalOffset'
import DashboardContent from './DashboardContent'
import EditBar from '../ControlBar/EditBar'
import NoContentMessage from '../../widgets/NoContentMessage'

export const Content = ({ updateAccess }) => {
    return updateAccess ? (
        <DashboardContent editMode={true} />
    ) : (
        <NoContentMessage text={i18n.t('No access')} />
    )
}

Content.propTypes = {
    updateAccess: PropTypes.bool,
}
export class EditDashboard extends Component {
    state = {
        initialized: false,
    }

    initEditDashboard = () => {
        if (this.props.dashboard) {
            this.setState({ initialized: true })
            this.props.setEditDashboard(this.props.dashboard, this.props.items)
        }
    }

    componentDidMount() {
        this.initEditDashboard()
    }

    componentDidUpdate() {
        if (!this.state.initialized) {
            this.initEditDashboard()
        }
    }

    getDashboardContent = () => {
        const contentNotReady =
            !this.props.dashboardsLoaded || this.props.id === null

        return (
            <div className="dashboard-wrapper">
                {contentNotReady ? null : (
                    <Content updateAccess={this.props.updateAccess} />
                )}
            </div>
        )
    }

    render() {
        return (
            <>
                <EditBar />
                <DashboardVerticalOffset editMode={true} />
                {this.getDashboardContent()}
            </>
        )
    }
}

EditDashboard.propTypes = {
    dashboard: PropTypes.object,
    dashboardsLoaded: PropTypes.bool,
    id: PropTypes.string,
    items: PropTypes.array,
    setEditDashboard: PropTypes.func,
    updateAccess: PropTypes.bool,
}

const mapStateToProps = state => {
    const id = sGetSelectedId(state)
    const dashboard = id ? sGetDashboardById(state, id) : null

    const updateAccess =
        dashboard && dashboard.access ? dashboard.access.update : false

    return {
        dashboard,
        id,
        updateAccess,
        items: sGetDashboardItems(state),
        dashboardsLoaded: !sDashboardsIsFetching(state),
    }
}

export default connect(mapStateToProps, {
    setEditDashboard: acSetEditDashboard,
})(EditDashboard)
