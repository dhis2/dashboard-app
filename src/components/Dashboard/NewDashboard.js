import React, { Component, Fragment } from 'react';
import { connect } from 'react-redux';

import { acSetEditNewDashboard } from '../../actions/editDashboard';
import DashboardVerticalOffset from './DashboardVerticalOffset';
import EditBar from '../ControlBar/EditBar';
import DashboardContent from './DashboardContent';

class NewDashboard extends Component {
    componentDidMount() {
        this.props.setNewDashboard();
    }

    render() {
        return (
            <Fragment>
                <EditBar />
                <DashboardVerticalOffset editMode={true} />
                <div className="dashboard-wrapper">
                    <DashboardContent editMode={true} />
                </div>
            </Fragment>
        );
    }
}

export default connect(
    null,
    {
        setNewDashboard: acSetEditNewDashboard,
    }
)(NewDashboard);
