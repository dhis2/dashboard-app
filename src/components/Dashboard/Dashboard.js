import React, { Component } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

import { tSelectDashboard } from '../../actions/dashboards'
import { sDashboardsIsFetching } from '../../reducers/dashboards'
import { EDIT, NEW } from './dashboardModes'
import ViewDashboard from './ViewDashboard'
import EditDashboard from './EditDashboard'
import NewDashboard from './NewDashboard'

class Dashboard extends Component {
    setDashboard = () => {
        if (this.props.dashboardsLoaded) {
            const id = this.props.match.params.dashboardId || null

            this.props.selectDashboard(id)

            this.scrollToTop()
        }
    }

    scrollToTop = () => {
        window.scrollTo(0, 0)
    }

    componentDidMount() {
        this.setDashboard()
    }

    componentDidUpdate() {
        this.setDashboard()
    }

    render() {
        switch (this.props.mode) {
            case EDIT:
                return <EditDashboard />
            case NEW:
                return <NewDashboard />
            default:
                return <ViewDashboard />
        }
    }
}

Dashboard.propTypes = {
    dashboardsLoaded: PropTypes.bool,
    match: PropTypes.object,
    mode: PropTypes.string,
    selectDashboard: PropTypes.func,
}

const mapStateToProps = state => {
    return { dashboardsLoaded: !sDashboardsIsFetching(state) }
}

export default connect(mapStateToProps, {
    selectDashboard: tSelectDashboard,
})(Dashboard)
