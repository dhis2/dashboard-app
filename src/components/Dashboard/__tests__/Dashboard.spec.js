import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { Dashboard } from '../Dashboard'
import { NEW, VIEW, EDIT, PRINT, PRINT_LAYOUT } from '../dashboardModes'
import { NON_EXISTING_DASHBOARD_ID } from '../../../reducers/selected'

jest.mock('../../ControlBar/DashboardsBar', () => 'DashboardsBar')
jest.mock('../DashboardVerticalOffset', () => 'DashboardVerticalOffset')
jest.mock('../../../widgets/NoContentMessage', () => 'NoContentMessage')
jest.mock('../ViewDashboard', () => 'ViewDashboard')
jest.mock('../EditDashboard', () => 'EditDashboard')
jest.mock('../NewDashboard', () => 'NewDashboard')
jest.mock('../PrintDashboard', () => 'PrintDashboard')
jest.mock('../PrintLayoutDashboard', () => 'PrintLayoutDashboard')

describe('Dashboard', () => {
    let props
    let shallowDashboard
    const dashboard = () => {
        if (!shallowDashboard) {
            shallowDashboard = shallow(<Dashboard {...props} />)
        }
        return shallowDashboard
    }

    Object.defineProperty(global.document, 'getElementsByTagName', {
        value: () => [
            {
                classList: {
                    add: Function.prototype,
                    remove: Function.prototype,
                },
            },
        ],
    })

    Object.defineProperty(global.window, 'scrollTo', {
        value: Function.prototype,
    })

    beforeEach(() => {
        props = {
            dashboardsIsEmpty: false,
            dashboardsLoaded: false,
            id: null,
            match: { params: { dashboardId: null } },
            mode: VIEW,
            selectDashboard: jest.fn(),
        }
        shallowDashboard = undefined
    })
    it('renders correctly when dashboards not loaded and id is null', () => {
        expect(toJson(dashboard())).toMatchSnapshot()
    })

    it('renders correctly when dashboards loaded and id is null', () => {
        props.dashboardsLoaded = true
        expect(toJson(dashboard())).toMatchSnapshot()
    })

    it('renders correctly when dashboards not loaded and id is not null', () => {
        props.id = 'rainbowdash'
        expect(toJson(dashboard())).toMatchSnapshot()
    })

    it('renders NEW dashboard', () => {
        props.dashboardsLoaded = true
        props.id = 'rainbowdash'
        props.mode = NEW
        expect(toJson(dashboard())).toMatchSnapshot()
    })

    it('renders correctly when no dashboards found', () => {
        props.dashboardsLoaded = true
        props.dashboardsIsEmpty = true
        props.id = null
        props.mode = VIEW
        expect(toJson(dashboard())).toMatchSnapshot()
    })

    it('renders correctly when dashboard id is not valid', () => {
        props.dashboardsLoaded = true
        props.id = NON_EXISTING_DASHBOARD_ID
        props.mode = VIEW
        expect(toJson(dashboard())).toMatchSnapshot()
    })

    it('renders EDIT dashboard', () => {
        props.dashboardsLoaded = true
        props.id = 'rainbowdash'
        props.mode = EDIT
        expect(toJson(dashboard())).toMatchSnapshot()
    })

    it('renders PRINT dashboard', () => {
        props.dashboardsLoaded = true
        props.id = 'rainbowdash'
        props.mode = PRINT
        expect(toJson(dashboard())).toMatchSnapshot()
    })

    it('renders PRINT_LAYOUT dashboard', () => {
        props.dashboardsLoaded = true
        props.id = 'rainbowdash'
        props.mode = PRINT_LAYOUT
        expect(toJson(dashboard())).toMatchSnapshot()
    })
})
