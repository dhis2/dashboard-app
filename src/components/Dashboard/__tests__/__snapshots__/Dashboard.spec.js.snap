// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard renders EDIT dashboard 1`] = `
<WindowDimensionsProvider>
  <Dashboard
    dashboardsIsEmpty={false}
    dashboardsLoaded={true}
    id="rainbowdash"
    match={
      Object {
        "params": Object {
          "dashboardId": null,
        },
      }
    }
    mode="edit"
    selectDashboard={
      [MockFunction] {
        "calls": Array [
          Array [
            undefined,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <EditDashboard />
  </Dashboard>
</WindowDimensionsProvider>
`;

exports[`Dashboard renders NEW dashboard 1`] = `
<WindowDimensionsProvider>
  <Dashboard
    dashboardsIsEmpty={false}
    dashboardsLoaded={true}
    id="rainbowdash"
    match={
      Object {
        "params": Object {
          "dashboardId": null,
        },
      }
    }
    mode="new"
    selectDashboard={
      [MockFunction] {
        "calls": Array [
          Array [
            undefined,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <NewDashboard />
  </Dashboard>
</WindowDimensionsProvider>
`;

exports[`Dashboard renders PRINT dashboard 1`] = `
<WindowDimensionsProvider>
  <Dashboard
    dashboardsIsEmpty={false}
    dashboardsLoaded={true}
    id="rainbowdash"
    match={
      Object {
        "params": Object {
          "dashboardId": null,
        },
      }
    }
    mode="print"
    selectDashboard={
      [MockFunction] {
        "calls": Array [
          Array [
            undefined,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <PrintDashboard />
  </Dashboard>
</WindowDimensionsProvider>
`;

exports[`Dashboard renders PRINT_LAYOUT dashboard 1`] = `
<WindowDimensionsProvider>
  <Dashboard
    dashboardsIsEmpty={false}
    dashboardsLoaded={true}
    id="rainbowdash"
    match={
      Object {
        "params": Object {
          "dashboardId": null,
        },
      }
    }
    mode="print-layout"
    selectDashboard={
      [MockFunction] {
        "calls": Array [
          Array [
            undefined,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <PrintLayoutDashboard />
  </Dashboard>
</WindowDimensionsProvider>
`;

exports[`Dashboard renders correctly when dashboard id is not valid 1`] = `
<WindowDimensionsProvider>
  <Dashboard
    dashboardsIsEmpty={false}
    dashboardsLoaded={true}
    id="0"
    match={
      Object {
        "params": Object {
          "dashboardId": null,
        },
      }
    }
    mode="view"
    selectDashboard={
      [MockFunction] {
        "calls": Array [
          Array [
            undefined,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <DashboardsBar />
    <NoContentMessage
      text="Requested dashboard not found"
    />
  </Dashboard>
</WindowDimensionsProvider>
`;

exports[`Dashboard renders correctly when dashboards loaded and id is null 1`] = `
<WindowDimensionsProvider>
  <Dashboard
    dashboardsIsEmpty={false}
    dashboardsLoaded={true}
    id={null}
    match={
      Object {
        "params": Object {
          "dashboardId": null,
        },
      }
    }
    mode="view"
    selectDashboard={
      [MockFunction] {
        "calls": Array [
          Array [
            undefined,
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
  >
    <Layer
      translucent={true}
    />
  </Dashboard>
</WindowDimensionsProvider>
`;

exports[`Dashboard renders correctly when dashboards not loaded and id is not null 1`] = `
<WindowDimensionsProvider>
  <Dashboard
    dashboardsIsEmpty={false}
    dashboardsLoaded={false}
    id="rainbowdash"
    match={
      Object {
        "params": Object {
          "dashboardId": null,
        },
      }
    }
    mode="view"
    selectDashboard={[MockFunction]}
  >
    <Layer
      translucent={true}
    />
  </Dashboard>
</WindowDimensionsProvider>
`;

exports[`Dashboard renders correctly when dashboards not loaded and id is null 1`] = `
<WindowDimensionsProvider>
  <Dashboard
    dashboardsIsEmpty={false}
    dashboardsLoaded={false}
    id={null}
    match={
      Object {
        "params": Object {
          "dashboardId": null,
        },
      }
    }
    mode="view"
    selectDashboard={[MockFunction]}
  >
    <Layer
      translucent={true}
    />
  </Dashboard>
</WindowDimensionsProvider>
`;

exports[`Dashboard renders correctly when no dashboards found 1`] = `
<WindowDimensionsProvider>
  <Dashboard
    dashboardsIsEmpty={true}
    dashboardsLoaded={true}
    id="rainbowdash"
    match={
      Object {
        "params": Object {
          "dashboardId": null,
        },
      }
    }
    mode="view"
    selectDashboard={[MockFunction]}
  >
    <DashboardsBar />
    <NoContentMessage
      text="No dashboards found. Use the + button to create a new dashboard."
    />
  </Dashboard>
</WindowDimensionsProvider>
`;
